### PR TITLE
Scala - Starter bot

### PR DESCRIPTION
**Pull request by niekz**
Added a basic Scala starter bot

### CODE REVIEW

**Comment by niekz** [comment]
I will do so and update the pull request

File: starter-bots/scala/bot.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "author": "Johann Venter",
+  "email": "johannventer2@gmail.com",
+  "nickName": "Johann",
+  "botLocation": "/target/scala-2.12",
+  "botFileName": "entelect-scala-bot_2.12-0.1.jar",
+  "botLanguage": "scala"
+}
```
Comment (Blaarkies) [inline starter-bots/scala/bot.json]: Starter bots will be used by anyone in the community to build their bots, we need to remove personal information from these.
Please change it to:
```
	"author": "John Doe",
	"email": "john.doe@example.com",
	"nickName": "Odersky",
```
we use the designer/creator of the language's surname as the nickname in the starter-bot, in recognition of them. 
P.S, just double check this is correct, Scala was created by "Martin Odersky", right?

File: game-runner/src/main/java/za/co/entelect/challenge/botrunners/BotRunnerFactory.java [hunk 2]
```diff
@@ -28,6 +28,8 @@ public static BotRunner createBotRunner(BotMetaData botMetaData, int timeoutInMi
                 return new HaskellBotRunner(botMetaData, timeoutInMilliseconds);
             case PHP:
                 return new PHPBotRunner(botMetaData, timeoutInMilliseconds);
+            case SCALA:
+                return new ScalaBotRunner(botMetaData, timeoutInMilliseconds);
```
Comment (Blaarkies) [inline game-runner/src/main/java/za/co/entelect/challenge/botrunners/BotRunnerFactory.java]: In the BotLanguage.java file, you need to add SCALA as a language type, with this line:
```
    @SerializedName("scala")
    SCALA,
```

**Comment by Blaarkies** [comment]
Nevermind the bot.json or botlanguage changes, i already made those locally and pushed them to your branch. Just remember to rebase or fast-forward your readme 😄

**Comment by niekz** [comment]
@Blaarkies Hi, thank you for making the changes 😄. I have expanded the readme.md to be more comprehensive

File: starter-bots/scala/src/main/scala/za/co/sample/App.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package za.co.sample
+
+import za.co.sample.models.Models.{Cell, GameDetails, Player, State}
+
+import scala.util.Random
+
+object App {
+  def main(args: Array[String]): Unit = {
+    val stateFile = "state.json"
+    val commandFile = "command.json"
```
Comment (Blaarkies) [inline starter-bots/scala/src/main/scala/za/co/sample/App.scala]: This should be txt instead of JSON
```
val commandFile = "command.txt"
```

**Comment by Blaarkies** [comment]
@niekz 
I am having trouble getting the entelect-scala-bot_2.12-0.1.jar to run, even from the cmd terminal. Any idea what this is about?
```
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.NoClassDefFoundError: scala/Product
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.getDeclaredMethods0(Native Method)
        at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
        at java.lang.Class.privateGetMethodRecursive(Class.java:3048)
        at java.lang.Class.getMethod0(Class.java:3018)
        at java.lang.Class.getMethod(Class.java:1784)
        at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
        at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
Caused by: java.lang.ClassNotFoundException: scala.Product
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 19 more
```

**Comment by niekz** [comment]
@Blaarkies 
Getting the same issue now, not quite sure whats wrong, as generally it would indicate a lib that is compile with the wrong scala version (not the one indicated in the build.sbt). I'm looking into it currently, but the versions line up with other projects that run correctly on my side.

**Comment by Blaarkies** [comment]
@niekz 
I compiled using your new changes, but I get this when running only your bot from cmd with the 
```java -jar Entelect-Scala-Bot-assembly-0.1.jar``` command
```
Exception in thread "main" java.io.FileNotFoundException: state.json (The system cannot find the file specified)
```
It dumps a bunch of json data as well, so it finds something. Im not sure how to check for current folder structure in scala, do you mind helping me fix this?

**Comment by niekz** [comment]
@Blaarkies 
I'm going to go through the process of debugging it tonight, as well as doing some testing to make sure everything is working as intended. It worked with the state.json supplied in the samples directory, but I might just be doing something silly in the code, as I tested it without packing as a JAR before submitting the pull request. Will keep you updated as I work through it

**Comment by niekz** [comment]
@Blaarkies 
I have made changes to the code to resolve the Exception that was happening, it was due to changes I made to reading in the file before submitting the pull request, and didn't fix initially. The jar now executes successfully with a state.json in its directory, and outputs a valid command.txt

**Comment by niekz** [comment]
Found some more issues. Testing locally for now, needs more updating before we can merge it in

**Comment by Blaarkies** [comment]
Apologies for the long delay, the tournament kept us quite busy.

The new changes work and i see no problem with the bot, except for the bot language defined as "java", but i will quickly fix that after the merge

**Comment by niekz** [comment]
The bot has a logic issue with detecting attacked rows, work has been
keeping me busy, but I will be working on the bot in the next couple of
weeks. The first couple of rounds start out fine, but it has an issue when
a row is full with detecting being attacked, thus doesn't follow the rules
for starter bots completely

**Pull request merged**